File: faster.py
import math
import timeit
import time

def calcDistance(ax, ay, bx, by):
    return math.sqrt(((ax-bx)**2)+((ay-by)**2))

def calcMidpoint(ax, ay, bx, by):
    return [(ax+bx)*0.5, (ay+by)*0.5]

def calcPi():
    bx, by = 1, 0
    totArea = 0.5
    for i in range(2, 30):
        midpoint = calcMidpoint(0, 1, bx, by)
        midpointX, midpointY = midpoint[0], midpoint[1]
        midDistance = calcDistance(0, 0, midpointX, midpointY)
        normalizedX = midpointX/midDistance
        normalizedY = midpointY/midDistance
        triHeight = 1-midDistance
        triWidth = calcDistance(0, 1, bx, by)
        triSize = (triHeight*triWidth)*0.5
        totArea += triSize*(2**(i-2))
        
        bx, by = normalizedX, normalizedY
    return totArea*4

def library():
    runNumber = int(input("How many runs would you like to have? "))
    execution_time = timeit.timeit("calcPi()", globals=globals(), number=runNumber)
    print(f"Pi = {calcPi()}")
    print(f"Total Time Taken: {execution_time} Seconds")
    print(f"Average execution time over {runNumber} runs: {(execution_time/runNumber)*1000000:.5f} Microseconds")

def manual():
    runNumber = int(input("How many runs would you like to have? "))
    startT = time.time()
    for i in range(runNumber):
        calcPi()
    endT = time.time()
    print(f"Pi = {calcPi()}")
    print(f"Total Time Taken: {(endT-startT)} Seconds")
    print(f"Average execution time over {runNumber} runs: {((endT-startT)/runNumber)*1000000:.5f} Microseconds")
    
def main():
    method = int(input("How would you like to measure time (1/2)\n    1. Library (timeit)\n    2. Manual\n"))
    if method == 1:
        library()
    else:
        manual()

main()

File: faster.c
#include <stdio.h>
#include <math.h>
#include <time.h>
#include <locale.h>

long double calcDistance (long double ax, long double ay, long double bx, long double by) {
    return sqrt(((ax-bx)*(ax-bx))+((ay-by)*(ay-by)));
}

// Not the most efficient but I don't have internet so I can't search up how to return array
long double midpointX (long double ax, long double bx) {
    return (ax+bx)*0.5;
}

long double midpointY (long double ay, long double by) {
    return (ay+by)*0.5;
}

long double calcPi () {
    int size = 1;
    // Sets starting point
    long double bx = size;
    long double by = 0;
    long double totArea = size*size*0.5;
    for (int i = 0; i < 30; i++) {
        // Finds midpoint
        long double midX = midpointX(0, bx);
        long double midY = midpointY(size, by);
        // Normalizes to be on circle
        long double midDistance = calcDistance(0, 0, midX, midY);
        long double normalizedY = (midY/midDistance)*size;
        long double normalizedX = (midX/midDistance)*size;
        // Calculates triangle size
        long double triHeight = size-midDistance;
        long double triWidth = calcDistance(0, size, bx, by);
        long double triSize = (triHeight*triWidth)*0.5;
        // Multiplies by 2^i because each time, there are double the amount of each triangle
        totArea += triSize*(pow(2, (i)));
        // Sets the point to the normalized point
        bx = normalizedX;
        by = normalizedY;
    }
    // Multiplies by 4
    return (totArea*4)/(size*size);

}

int main() {
    setlocale(LC_NUMERIC, "");                  // For better formatting when printing
    // Number of runs
    int runNumber = 100000;
    clock_t start, end;                         // All the time stuff I found online
    double cpu_time_used;
    start = clock();
    
    // Main loop for time testing
    for (int i = 0; i < runNumber; i++) {
        calcPi();
    }

    end = clock();
    cpu_time_used = ((double) (end - start)) / CLOCKS_PER_SEC;

    printf("Pi: %.14Lf\n", calcPi());
    printf("Time Taken: %f Seconds for %'d runs\n", cpu_time_used, runNumber);
    float timePerRun = (cpu_time_used/runNumber)*1000000;
    printf("Average Time Taken: %f Microseconds\n", timePerRun);

    return 0;
} 